### DISCUSSION TITLE
Is federation supported (poly/micro services)?

### DISCUSSION THREAD

**Discussion by smolattack** [hayes/pothos#211] [Q&A] [answered]
This is the sort of thing I'm referring to: https://www.apollographql.com/docs/federation

Basically a way of stitching together disparate services that relate/link to each other (as opposed to disparate services that do not relate to each other, such as Cloudinary and G Maps Platform, even though you might use those two different services in the same app). I looked at the subgraph plugin but can't quite grok where it fits in

**Comment by hayes** [comment]
the sub-graph plugin is kinda the opposite, it lets you split out chunks of your graph into smaller pieces, the most common case for this is having an public API that is a subset of your internal API, using the same code base.

Federation is not something I have built out explicit support for.

I haven't looked closely at it, but federation v2 is something that probably could work pretty well with giraphql, but I don't have a plugin or any examples that show how it would work.

I am planning on building a plugin for it, but I haven't made much progress on it recently (#97).  With v2 being released, things get a lot easier.

GiraphQL is just a schema builder, so it wouldn't do any of the federation stuff itself, but I will eventually be writing a plugin that makes writing an schema compatible with apollo federation v2 easier

**Reply by smolattack** [reply]
Thanks for the detailed response. Apollo federation also uses the term subgraph so thanks for clearing that up. I'll keep an eye out for the plugin. Curious to see how paging and n+1 might be solved. I guess you wouldn't be able to use the Prisma dataloader helpers as you'd have multiple Prisma schemas/dbs.

**Reply by hayes** [reply]
I think wundergraph implements a data-loader pattern 🤷 would be outside the scope of giraphql, but federated graphs can definitely still implement optimizations like dataloaders

**Reply by smolattack** [reply]
OK that makes sense. I just read up more about federation and it seems Apollo makes sure it doesn't spam your services. You just make sure that your `__resolveReference` function uses a dataloader pattern.

Thanks for answering all my stupid questions, really helps!

**Comment by hayes** [comment]
Work on a plugin will mostly be happening in the PR: https://github.com/hayes/giraphql/pull/212

**Comment by hayes** [comment, marked as answer]
There is now a federation plugin!  This is an initial release, and is still pre-1.0 because federation 2 is still in alpha, but hopefully this will work for anyone trying to figure out federation with Pothos

https://www.pothos-graphql.dev/docs/plugins/federation
